app/employee-app/backend/modules/database/db_functions.bal

```diff
@@ -0,0 +1,443 @@
+#
+# + filter - Filter to retrieve the compensation data
+# + return - Compensation data or error
+public isolated function getCompensation(string filter) returns types:CompensationEmail|error {
+
+    CompensationEmail|error compensationRecord = databaseClient->queryRow(getCompensationQuery(filter));
+    if compensationRecord is error {
+        if compensationRecord is sql:Error && compensationRecord is sql:NoRowsError {
+            return error(string `No matching compensation data found for ${filter}!`);
+        }
+        string customError = string `Error occurred while retrieving the compensation data for ${filter}!`;
+        log:printError(customError, compensationRecord);
+        return error(customError);
+    }
+
+    types:Compensation[]|error compensation = compensationRecord.keyValuePairs.fromJsonStringWithType();
+    if compensation is error {
+        string customError = string `Error occurred while converting compensation data for ${filter}!`;
+        log:printError(customError, compensation);
+        return error(customError);
+    }
+    if (compensation.length() == 0) {
+        error customError = error(string `No matching compensation data found for ${filter}`);
+        return customError;
+    }
+
+    return {emailTemplate: compensationRecord.emailTemplate, compensation};
+}
+
+# Retrieve companies under from HRIS.
+#
+# + filter - Criteria to filter the data  
+# + 'limit - Number of records to retrieve  
+# + offset - Number of records to offset
+# + return - List of companies
+public isolated function getCompanies(CompanyFilter filter, int 'limit, int offset)
+    returns Company[]|error {
+
+    stream<DBCompany, sql:Error?> resultStream = databaseClient->query(getCompaniesQuery(filter, 'limit, offset));
+    Company[] companies = [];
+    error? iterateError = from DBCompany company in resultStream
+        do {
+            Office[]|error offices = company.offices.fromJsonStringWithType();
+            if offices is error {
+                string errorMsg = string `An error occurred when retrieving offices data of ${company.company} !`;
+                log:printError(errorMsg, offices);
+                return error(errorMsg);
+            }
+            companies.push({
+                id: company.id,
+                company: company.company,
+                location: company.location,
+                offices: offices
+            });
+        };
+    if iterateError is sql:Error {
+        if iterateError is sql:NoRowsError {
+            return [];
+        }
+        string errorMsg = string `An error occurred when retrieving companies!`;
+        log:printError(errorMsg, iterateError);
+        return error(errorMsg);
+    }
+    return companies;
+}
+
+# Get basic information about a given active employee.
+#
+# + email - Email of the employee
+# + return - Returns the basic information of the employee or an error
+public isolated function getEmployee(string email) returns Employee|error? {
```

Comment:
the return type includes `error`, but we don't effectively return any errors within the function. so can we make the return type `Employee?`  ?

Reply:
Added the returning error message.